quickwit/quickwit-storage/src/timeout_and_retry_storage.rs

```diff
@@ -0,0 +1,270 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::ops::Range;
+use std::path::Path;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use quickwit_common::rate_limited_info;
+use quickwit_common::uri::Uri;
+use quickwit_config::StorageTimeoutPolicy;
+use tantivy::directory::OwnedBytes;
+use tokio::io::AsyncRead;
+
+use crate::storage::SendableAsync;
+use crate::{BulkDeleteError, PutPayload, Storage, StorageErrorKind, StorageResult};
+
+/// Storage proxy that implements a retry operation if the underlying storage
+/// takes too long.
+///
+/// This is useful in order to unsure a low latency on S3.
+/// Retrying agressively is recommended for S3.
+/// https://docs.aws.amazon.com/whitepapers/latest/s3-optimizing-performance-best-practices/timeouts-and-retries-for-latency-sensitive-applications.html
+#[derive(Clone, Debug)]
+pub struct TimeoutAndRetryStorage {
+    underlying: Arc<dyn Storage>,
+    storage_timeout_policy: StorageTimeoutPolicy,
+}
+
+impl TimeoutAndRetryStorage {
+    /// Creates a new `TimeoutAndRetryStorage`.
+    ///
+    /// See [StorageTimeoutPolicy] for more information.
+    pub fn new(storage: Arc<dyn Storage>, storage_timeout_policy: StorageTimeoutPolicy) -> Self {
+        TimeoutAndRetryStorage {
+            underlying: storage,
+            storage_timeout_policy,
+        }
+    }
+}
+
+#[async_trait]
+impl Storage for TimeoutAndRetryStorage {
+    async fn check_connectivity(&self) -> anyhow::Result<()> {
+        self.underlying.check_connectivity().await
+    }
+
+    async fn put(&self, path: &Path, payload: Box<dyn PutPayload>) -> StorageResult<()> {
+        self.underlying.put(path, payload).await
+    }
+
+    fn copy_to<'life0, 'life1, 'life2, 'async_trait>(
+        &'life0 self,
+        path: &'life1 Path,
+        output: &'life2 mut dyn SendableAsync,
+    ) -> ::core::pin::Pin<
+        Box<
+            dyn ::core::future::Future<Output = StorageResult<()>>
+                + ::core::marker::Send
+                + 'async_trait,
+        >,
+    >
+    where
+        'life0: 'async_trait,
+        'life1: 'async_trait,
+        'life2: 'async_trait,
+        Self: 'async_trait,
+    {
+        self.underlying.copy_to(path, output)
+    }
+
+    async fn copy_to_file(&self, path: &Path, output_path: &Path) -> StorageResult<u64> {
+        self.underlying.copy_to_file(path, output_path).await
+    }
+
+    /// Downloads a slice of a file from the storage, and returns an in memory buffer
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> StorageResult<OwnedBytes> {
+        let num_bytes = range.len();
+        for (attempt_id, timeout_duration) in self
```

Comment:
One concern I would have here is that this will likely chew into rate limits much faster.
I think we're pretty well with in the 5500/s range. But we do see spikes / surges from the searcher upwards of 3K+/s currently.

Are storage failures recorded in a metric or logged so that we can identify failures / rate limits?

Reply:
The rate limiting is per prefix and overall.

We are working on improving your number of splits per day considerably.
This should improve your figure by a lot.

We were also planning to add a random prefix to split's ULID to avoid creating some hotspot... Unfortunately, your one-index-per-day scheme will make it useless.

Reply:
`Does the quickwit_storage_object_storage_gets_total metric also get incremented on a retry?

yes

`Are storage failures recorded in a metric or logged so that we can identify failures / rate limits?`

Storage timeout have a metric yes. Rate limiting will show up in logs.

Reply:
How is an error like that handled? Does the entire request fail, or does it surface as a failed shard with the recent change you made?

Reply:
There will be one retry happening on a different node. If that retry fails too, then the split will show up as a failed shards.